README.md
# NimbleTCodeSerial

Firmware that implements the [Toy Code (T-Code) v0.3](https://github.com/multiaxis/tcode-spec) protocol over USB/Serial Port for the [Nimble Connectivity Module](https://shop.exploratorydevices.com/product/connectivity-module-dev-kit/), an ESP32 controller for the [NimbleStroker](https://shop.exploratorydevices.com/).

Tested with Intiface Central (using custom configuration) and MultiFunPlayer's Serial output type, which connects to the Serial COM port interface of the NimbleConModule and sends T-Code commands as messages to the device. The T-Code commands are converted into position values which are then sent to the NimbleStroker actuator.

## TCode Information

- `D0` - Identity device and firmware version: `NimbleStroker_TCode_Serial_v0.3`
- `D1` - Identify TCode version: `TCode v0.4`
- `D2` - List available axes and user range preferences:
  - `L0 0 9999 Up`: **Up/down position** linear motion (default: `5000`)
    - Maps to NimbleStroker positions: -750 to 750
  - `V0 0 9999 Vibe`: **Vibration intensity** (default: `0`)
    - An oscillation is applied to the position when sent to the NimbleStroker.
    - Maps to the amplitude of oscillation: 0 to 25 (position units)
  - `A0 0 9999 Air`: **Auxilliary air in/out valve** (default `5000`)
    - Value `0000` = air-out valve (looser)
    - Value `5000` = stop valve (default)
    - Value `9999` = air-in valve (tighter)
  - `A1 0 9999 Force`: **Force command** (default `9999`)
    - Maps to NimbleStroker force command values: 0 to 1023
    - Controls the air pressure force of the actuator(?)
  - `A2 0 9999 VibSpeed`: **Vibration speed** (default: `9999`)
    - Maps to an oscillation speed for vibration: 0 to 20hz (default 20hz)

Other info:

- Sending live control values to an axis will ease to the target value over multiple frames rather than jump immediately when the difference in change is large (> 100 t-code units, or >50 position units). This is intended to protect the user and device. ([Source1](https://github.com/mnh86/NimbleTCodeSerial/blob/6ab66638b2670115e770fdee9d2ec5c7b04f9390/include/TCodeAxis.h#L217-L228), [Source2](https://github.com/mnh86/NimbleTCodeSerial/blob/6ab66638b2670115e770fdee9d2ec5c7b04f9390/src/main.cpp#L104-L111))
- Up/down position axis values that are sent to the NimbleStroker are set as (-750 to 750) instead of the full [documented range of (-1000 to 1000)](https://github.com/ExploratoryDevices/NimbleConModule/blob/31f09fbcaa068b3d7fe8d47e44ea5ed11437c852/README.md?plain=1#L30) to avoid piston damaging the actuator (slamming occurs at min/max ranges). This aligns with the same max/min values that the NimbleStroker Pendant sends to the actuator, from debug log analysis.
- [Acutuator feedback values](https://github.com/ExploratoryDevices/NimbleConModule/blob/31f09fbcaa068b3d7fe8d47e44ea5ed11437c852/README.md?plain=1#L24-L27) are not currently exposed in this firmware. Possible options could be to extend the tcode interface to support sensor data, or implement a separate interface (such as websockets) that clients could connect on.

## Usage

1. Install [Windows Virtual COM Port (VCP) drivers](https://github.com/mnh86/NimbleConModule/blob/feat/docs/docs/setup-guide-windows-arduino-ide1.md#install-windows-virtual-com-port-vcp-drivers) for the USB/serial connection to the module.
2. Set up [VSCode with PlatformIO](https://randomnerdtutorials.com/vs-code-platformio-ide-esp32-esp8266-arduino/)
3. Clone this repo and open the project in VSCode
4. Build and upload this program into the NimbleConModule (USB/serial)
5. Attach the NimbleConModule to the actuator (Label A)
   - Note: Pendant connection not supported
6. Open the PlatformIO Serial Monitor. Enter a TCode command (ie. `D2`) to test.
7. Click the Encoder Dial to toggle stop/start sending commands to the actuator.

## Testing with Intiface® Central

On Windows with [Intiface Central](https://intiface.com/central/) installed...

1. Open the Intiface config file: `C:\Users\(User)\AppData\Roaming\com.nonpolynomial\intiface_central\config\buttplug-device-config.json`
2. Find the [JSON block for `tcode-v03`](https://github.com/buttplugio/buttplug/blob/9159c402d866161d55363aad23626df6c006e518/buttplug/buttplug-device-config/buttplug-device-config.json#L4183-L4208). Change the `port` to the one with your attached NimbleConModule, and add a `ScalarCmd` for the vibration parameter. ie:
   ```
   ...
   "tcode-v03": {
      "serial": [
        {
          "port": "COM3",
          "baud-rate": 115200,
          "data-bits": 8,
          "parity": "N",
          "stop-bits": 1
        }
      ],
      "defaults": {
        "name": "NimbleStroker Connectivity Module (Serial TCode v0.3)",
        "messages": {
          "LinearCmd": [
            {
              "StepRange": [
                0,
                100
              ],
              "ActuatorType": "Position",
              "FeatureDescriptor": "Up/Down Position (L0)"
            }
          ],
          "ScalarCmd": [
            {
              "StepRange": [
                0,
                100
              ],
              "ActuatorType": "Vibrate",
              "FeatureDescriptor": "Vibration (V0)"
            }
          ]
        }
      }
    },
    ...
    ```
3. Launch the [Intiface Central](https://intiface.com/central/) application.
4. Under `Settings -> Device Managers`, toggle on `Serial Port`.
5. Click the `Start Server` Icon (top left) to start the server.
6. Under `Devices` click `Start Scanning`...
7. A new device should show up with the name configured in the JSON file.
8. Click the `Toggle Oscillation` button and watch the NimbleConModule LEDs spin.

![Buttplug Device UI](./docs/ButtplugIO-device-ui.jpg)

## Testing with MultiFunPlayer

1. Launch the [MultiFunPlayer](https://github.com/Yoooi0/MultiFunPlayer) application.
2. In the Output section, add a "Serial" device (plus sign).
3. Set the serial port to your COM port value (ie. "COM3").
4. Test the connection with the play button.
5. (Optional) To configure additional axes (L0, V0, A0, A1, A2) for multi script support...
6. Open the Output Configuration panel.
7. Clone the `TCode-0.3 (default)` config, name it `NimbleTcodeSerial`.
8. Enable the channels you'd like to use. ie.:
   - `L0` Up/Down (Select "Load unnamed script") - Default value: `50%`
   - `V0` Vibrate - Default value: `0%` (off state)
   - `A0` Valve - Default value: `50%` (off state)
   - `A1` Force (can rename) - Default value: `100%` (max)
   - `A2` Vibspeed (can rename) - Default value: `100%` (max)
9. See screenshots below.

![MultiFunPlayer Device Config](./docs/MFP-device-config.jpg)

![MultiFunPlayer Device UI](./docs/MFP-device-ui.jpg)

## Attributions

- Utilized examples from: <https://github.com/tyrm/nimblestroker/>
- Modified NimbleConSDK from: <https://github.com/ExploratoryDevices/NimbleConModule>
- Uses TCode library from: https://github.com/Dreamer2345/Arduino_TCode_Parser
- See also [platformio.ini](./platformio.ini) for other 3rd party OSS libraries used in this project
- @qdot on the [Buttplug.io Discord](https://discord.gg/h28chsBD) for support


CHANGELOG.md
# Change log

## v0.5 - 02/28/2023
- Change: Single click toggle will also reset the actuator state when stopped (position = 0, force = max, vibration = off)

## v0.4 - 02/21/2023
- Refactor TCode setup into NimbleTCode.h for reuse with different connection methods (Serial, Wifi, BTLE).
- Fix: Air in/out valve was getting stuck when device in off state (encoder press).
- Change: Single clicking the encoder now toggles the run/stop state for the device (instead of double click/long press).

## v0.3 - 02/06/2023
- Fix: Divide by zero crash during vibration calculations.
- `L0` Up Position: mapped range was changed from (-1000 -> 1000) to (-750 -> 750). Based on analysis of Pendant behavior which does not send values higher than 750. Testing values at 1000 causes slamming of the actuator piston to occur against its enclosure.
- `V0` Vibration Amplitude: internal range was changed from (0 -> 20) to (0 -> 25) position units. Its default is now 25, which matches more closely with the NimbleStroker Pendant behavior.
- `V0` Vibration Amplitude: changes will now ease in and out rather than a linear change.
- `A2` Vibration Speed: default was changed from 10hz to 20hz.

## v0.2 - 02/05/2023
- Added new aux axis for controlling vibration speed via tcode
  - `A2 0 9999 VibSpeed`: **Vibration speed** (default: `5000`)
  - Maps to an oscillation speed for vibration: 0 to 20.0hz (default 10.0hz)
- Changed Encoder button behavior:
  - `Long press (2sec)`: Stop sending commands to the actuator
  - `Double click`: Resume sending commands to the actuator
- Changed LED behavior so that vibration is more visible
- Added vibration control to buttplug-device.config.json example in README
- Removed TSTOP when tempLimiting is high, but added logging when it occurs

## v0.1 - 02/04/2023
- Initial PoC release
- RFC on axes spec


src/main.cpp
#include <Arduino.h>
#include <millisDelay.h>
#include <BfButton.h>
#include "NimbleTCode.h"

#define FIRMWAREVERSION "NimbleStroker_TCode_Serial_v0.4"

NimbleTCode nimble(FIRMWAREVERSION);

millisDelay ledUpdateDelay;
millisDelay logDelay;

BfButton btn(BfButton::STANDALONE_DIGITAL, ENC_BUTT, true, LOW);

void pressHandler(BfButton *btn, BfButton::press_pattern_t pattern)
{
    switch (pattern)
    {
    case BfButton::SINGLE_PRESS:
        nimble.toggle();
        if (!nimble.isRunning()) nimble.resetState();
        break;
    }
}

void logTimer()
{
    if (!logDelay.justFinished()) return;
    logDelay.repeat();

    nimble.printFrameState();
}

void updateLEDs()
{
    if (!ledUpdateDelay.justFinished()) return;
    ledUpdateDelay.repeat();

    nimble.updateEncoderLEDs();
    nimble.updateHardwareLEDs();
    nimble.updateNetworkLEDs(0, 0);
}

void setup()
{
    // NimbleStroker TCode setup
    nimble.init();
    while (!Serial);

#ifdef DEBUG
    Serial.setDebugOutput(true);
    delay(200);
    Serial.println("\nStarting");
#else
    Serial.setDebugOutput(false);
#endif

    // Button interface
    btn.onPress(pressHandler);

    // Timers setup
    ledUpdateDelay.start(30);
    logDelay.start(1000);
}

void loop()
{
    btn.read();
    while (Serial.available() > 0) {
        nimble.inputByte(Serial.read());
    }
    nimble.updateActuator();
    updateLEDs();
#ifdef DEBUG
    logTimer();
#endif
}

include/TCodeFixed.h
// TCode-Fixed-Point-H v1.0,
// protocal by TempestMAx (https://www.patreon.com/tempestvr)
// implemented by Eve 05/02/2022
// usage of this class can be found at (https://github.com/Dreamer2345/Arduino_TCode_Parser)
// Please copy, share, learn, innovate, give attribution.
// Q16 Implementation for fractional calculations on hardware without an FPU
// History:
//
#pragma once
#ifndef TCODE_FIXED_POINT_H
#define TCODE_FIXED_POINT_H
#include "Arduino.h"

#define Q 16

/*
int32_t sat_Q16(int64_t x)
{
	if (x > 0x7FFFFFFF) return 0x7FFFFFFF;
	else if (x < -0x80000000) return -0x80000000;
	else return (int32_t)x;
}
*/

int32_t addQ16(int32_t a, int32_t b)
{
	int64_t tmp = (int64_t)a + (int64_t)b;
	return tmp;
}

int32_t subQ16(int32_t a, int32_t b)
{
	return a - b;
}

int32_t multQ16(int32_t a, int32_t b)
{
	int64_t tmp = ((int64_t)a * (int64_t)b) >> Q;
	return tmp;
}

int32_t divQ16(int32_t a, int32_t b)
{
	return ((((int64_t)a) << Q) / b);
}

int32_t constrainQ16(int32_t v, int32_t min, int32_t max)
{
	if (v < min)
		return min;
	if (v > max)
		return max;
	return v;
}

int32_t Q16fromInt(int i)
{
	return ((int32_t)i) << Q;
}

int32_t Q16fromFloat(float i)
{
	float newValuef = i * (1 << Q);
	return (int32_t)newValuef;
}

int32_t Q16fromDouble(double i)
{
	double newValued = i * (1 << Q);
	return (int32_t)newValued;
}

int IntfromQ16(int32_t i)
{
	i = addQ16(i, Q16fromFloat(0.5f));
	return (int)(i >> Q);
}

float FloatfromQ16(int32_t i)
{
	return (((float)i) / (1 << Q));
}

double DoublefromQ16(int32_t i)
{
	return (((double)i) / (1 << Q));
}

#endif

include/TCodeBuffer.h
// TCode-Buffer-class v1.0,
// protocal by TempestMAx (https://www.patreon.com/tempestvr)
// implemented by Eve 04/06/2022
// usage of this class can be found at (https://github.com/Dreamer2345/Arduino_TCode_Parser)
// Please copy, share, learn, innovate, give attribution.
// Decodes T-code commands
// It can handle:
//   x linear channels (L0, L1, L2... L9)
//   x rotation channels (R0, R1, R2... L9)
//   x vibration channels (V0, V1, V2... V9)
//   x auxilliary channels (A0, A1, A2... A9)
// History:
//

#pragma once
#ifndef TCODE_BUFFER_H
#define TCODE_BUFFER_H

template <unsigned TCODE_BUFFER_LENGTH = 127>
class TCodeBuffer
{
private:
    unsigned Front = 0;
    unsigned Back = 0;
    char buffer[TCODE_BUFFER_LENGTH];

public:
    bool isFull();
    bool isEmpty();
    void clear();

    unsigned count();

    bool push(char obj);
    char pop(bool &success);
    char peek(bool &success);
    char pop();
    char peek();
};

template <unsigned TCODE_BUFFER_LENGTH>
bool TCodeBuffer<TCODE_BUFFER_LENGTH>::isEmpty()
{
    return Front == Back;
}

template <unsigned TCODE_BUFFER_LENGTH>
void TCodeBuffer<TCODE_BUFFER_LENGTH>::clear()
{
    Front = Back;
}

template <unsigned TCODE_BUFFER_LENGTH>
bool TCodeBuffer<TCODE_BUFFER_LENGTH>::isFull()
{
    return (Front == ((Back + 1) % TCODE_BUFFER_LENGTH));
}

template <unsigned TCODE_BUFFER_LENGTH>
unsigned TCodeBuffer<TCODE_BUFFER_LENGTH>::count()
{
    if (Back >= Front)
        return Back - Front;

    return TCODE_BUFFER_LENGTH - (Front - Back);
}

template <unsigned TCODE_BUFFER_LENGTH>
bool TCodeBuffer<TCODE_BUFFER_LENGTH>::push(char obj)
{
    if (isFull())
    {
        return false;
    }
    buffer[Back] = obj;
    Back = (Back + 1) % TCODE_BUFFER_LENGTH;
    return true;
}

template <unsigned TCODE_BUFFER_LENGTH>
char TCodeBuffer<TCODE_BUFFER_LENGTH>::pop(bool &success)
{
    if (isEmpty())
    {
        success = false;
        return 0;
    }

    success = true;
    char headItem = buffer[Front];
    Front = (Front + 1) % TCODE_BUFFER_LENGTH;
    return headItem;
}

template <unsigned TCODE_BUFFER_LENGTH>
char TCodeBuffer<TCODE_BUFFER_LENGTH>::peek(bool &success)
{
    if (isEmpty())
    {
        success = false;
        return 0;
    }

    success = true;
    char headItem = buffer[Front];
    return headItem;
}

template <unsigned TCODE_BUFFER_LENGTH>
char TCodeBuffer<TCODE_BUFFER_LENGTH>::pop()
{
    if (isEmpty())
    {
        return 0;
    }

    char headItem = buffer[Front];
    Front = (Front + 1) % TCODE_BUFFER_LENGTH;
    return headItem;
}

template <unsigned TCODE_BUFFER_LENGTH>
char TCodeBuffer<TCODE_BUFFER_LENGTH>::peek()
{
    if (isEmpty())
    {
        return 0;
    }

    char headItem = buffer[Front];
    return headItem;
}

#endif

include/nimbleConModule.h
#include <Arduino.h>
// From https://github.com/ExploratoryDevices/NimbleConModule (with edits)
#include <ESP32Encoder.h>   // https://github.com/madhephaestus/ESP32Encoder
#include <HardwareSerial.h> // Arduino Core ESP32 Hardware serial library

// min() function needs this to work on ESP32
#ifndef min
#define min(a, b) (((a) < (b)) ? (a) : (b))
#endif

// Encoder pins
#define ENC_BUTT 2
#define ENC_A 35
#define ENC_B 34

ESP32Encoder encoder;

// Serial pins
#define PEND_RX 14
#define PEND_TX 15
#define ACT_RX 16
#define ACT_TX 17

#define SERIAL_BAUD 115200

HardwareSerial pendSerial(1);
HardwareSerial actSerial(2);

#define PACKET_TIMEOUT 50 // Time duration (ms) for packet timeout

// ADC Pins
#define ADC_REF 32
#define SENSOR_ADC 33

// Encoder LED PWM channels
#define ENC_LED_E 0
#define ENC_LED_SE 1
#define ENC_LED_S 2
#define ENC_LED_SW 3
#define ENC_LED_W 4
#define ENC_LED_NW 5
#define ENC_LED_N 6
#define ENC_LED_NE 7

// Other LED PWM channels
#define ACT_LED 8
#define PEND_LED 9
#define BT_LED 10
#define WIFI_LED 11

#define LED_MAX_DUTY 75 // No visible brightness difference with values between 75 and 255.

// Timers for sending serial data to actuator and pendant
#define SEND_INTERVAL 2000 // microseconds between packets sent.

int timeSinceLastActSend = 0;
int timeSinceLastPendSend = 0;

volatile int timerTriggered;

hw_timer_t *timer = NULL;
portMUX_TYPE timerMux = portMUX_INITIALIZER_UNLOCKED;

void IRAM_ATTR onTimer()
{
    portENTER_CRITICAL_ISR(&timerMux);
    timerTriggered = 1; // Set timer as triggered.
    portEXIT_CRITICAL_ISR(&timerMux);
}

bool checkTimer()
{
    if (timerTriggered == 1)
    {
        portENTER_CRITICAL(&timerMux);
        timerTriggered = 0; // Clear timer flag
        portEXIT_CRITICAL(&timerMux);
        return (1); // Return 1 to indicate timer has triggered
    }
    return (0);
}

// Pendant Variables
#define IDLE_FORCE 200 // Centering force to send when no value position signal is received.
#define MAX_FORCE 1023 // Pendant uses this value as a constant

struct Pendant
{
    bool present;
    // Signals from the pendant
    long positionCommand; // (range: -1000 to 1000)
    long forceCommand;
    bool activated;
    bool airOut;
    bool airIn;

    // Signals to the pendant
    long positionFeedback; // (range: -1000 to 1000)
    long forceFeedback;    // (range: -1023 to 1023)
    bool tempLimiting;     // Set high if the actuator is thermally limiting its performance.
    bool sensorFault;      // Set high if there's a fault in the position sensor.
};

struct Pendant pendant; // Declare pendant

// Actuator Variables
#define ACTUATOR_MAX_POS 750

struct Actuator
{
    bool present;

    // Signals from the actuator
    long positionFeedback; // (range: -1000 to 1000)
                           // Errata: actuator delivered before January 2023 this signal always reads positive.
    long forceFeedback;    // (range: -1023 to 1023)
    bool tempLimiting;     // Reads high if the actuator is thermally limiting its performance.
    bool sensorFault;      // Reads high if there's a fault in the position sensor.

    // Signals to the actuator
    long positionCommand; // (range: -1000 to 1000)
    long forceCommand;    // (range: 0 to 1023)
    bool activated;       // Not used
    bool airOut;          // Set high to open air-out valve (looser)
    bool airIn;           // Set high to open air-in valve  (tighter)
};

struct Actuator actuator; // Declare actuator

// Initialization fuction
void initNimbleConModule()
{
    // Setup encoder
    pinMode(ENC_BUTT, INPUT_PULLUP);
    pinMode(ENC_A, INPUT_PULLUP);
    pinMode(ENC_B, INPUT_PULLUP);

    ESP32Encoder::useInternalWeakPullResistors = UP;
    encoder.attachHalfQuad(ENC_A, ENC_B);
    encoder.setCount(0);

    // Actuator/Pendant defaults
    pendant.forceCommand = IDLE_FORCE;
    actuator.forceCommand = IDLE_FORCE;

    // Setup serial ports
    Serial.begin(SERIAL_BAUD);                                   // open serial port for USB connection
    pendSerial.begin(SERIAL_BAUD, SERIAL_8N1, PEND_RX, PEND_TX); // open serial port for pendant
    actSerial.begin(SERIAL_BAUD, SERIAL_8N1, ACT_RX, ACT_TX);    // open serial port for actuator

    // Set up timer interrupt.
    timer = timerBegin(0, 80, true);             // Set timer parameters
    timerAttachInterrupt(timer, &onTimer, true); // Attach interrupt to ISR
    timerAlarmWrite(timer, SEND_INTERVAL, true); // Configure timer threshold
    timerAlarmEnable(timer);                     // Enable timer

    // Attach PWM to LED pins (pin, PWM channel)
    ledcAttachPin(4, ENC_LED_E);
    ledcAttachPin(5, ENC_LED_SE);
    ledcAttachPin(12, ENC_LED_S);
    ledcAttachPin(13, ENC_LED_SW);
    ledcAttachPin(21, ENC_LED_W);
    ledcAttachPin(22, ENC_LED_NW);
    ledcAttachPin(23, ENC_LED_N);
    ledcAttachPin(25, ENC_LED_NE);
    ledcAttachPin(18, ACT_LED);
    ledcAttachPin(19, PEND_LED);
    ledcAttachPin(26, BT_LED);
    ledcAttachPin(27, WIFI_LED);

    // Configure PWM channels (PWM channel, PWM frequency, PWM counter bits)
    ledcSetup(ENC_LED_E, 1000, 8);
    ledcSetup(ENC_LED_SE, 1000, 8);
    ledcSetup(ENC_LED_S, 1000, 8);
    ledcSetup(ENC_LED_SW, 1000, 8);
    ledcSetup(ENC_LED_W, 1000, 8);
    ledcSetup(ENC_LED_NW, 1000, 8);
    ledcSetup(ENC_LED_N, 1000, 8);
    ledcSetup(ENC_LED_NE, 1000, 8);
    ledcSetup(ACT_LED, 1000, 8);
    ledcSetup(PEND_LED, 1000, 8);
    ledcSetup(BT_LED, 1000, 8);
    ledcSetup(WIFI_LED, 1000, 8);
}

void ledLevelDisplay(byte LEDScale)
{
    LEDScale > 0 ? ledcWrite(ENC_LED_N, map(LEDScale, 0, 32, 0, LED_MAX_DUTY)) : ledcWrite(ENC_LED_N, 0);
    LEDScale > 32 ? ledcWrite(ENC_LED_NE, map(LEDScale, 32, 64, 0, LED_MAX_DUTY)) : ledcWrite(ENC_LED_NE, 0);
    LEDScale > 64 ? ledcWrite(ENC_LED_E, map(LEDScale, 64, 96, 0, LED_MAX_DUTY)) : ledcWrite(ENC_LED_E, 0);
    LEDScale > 96 ? ledcWrite(ENC_LED_SE, map(LEDScale, 96, 128, 0, LED_MAX_DUTY)) : ledcWrite(ENC_LED_SE, 0);
    LEDScale > 128 ? ledcWrite(ENC_LED_S, map(LEDScale, 128, 160, 0, LED_MAX_DUTY)) : ledcWrite(ENC_LED_S, 0);
    LEDScale > 160 ? ledcWrite(ENC_LED_SW, map(LEDScale, 160, 192, 0, LED_MAX_DUTY)) : ledcWrite(ENC_LED_SW, 0);
    LEDScale > 192 ? ledcWrite(ENC_LED_W, map(LEDScale, 192, 224, 0, LED_MAX_DUTY)) : ledcWrite(ENC_LED_W, 0);
    LEDScale > 224 ? ledcWrite(ENC_LED_NW, map(LEDScale, 224, 255, 0, LED_MAX_DUTY)) : ledcWrite(ENC_LED_NW, 0);
}

void ledPositionPulse(short int position, bool isOn)
{
    byte ledScale = map(abs(position), 1, ACTUATOR_MAX_POS, 1, LED_MAX_DUTY);
    byte ledState1 = 0;
    byte ledState2 = 0;

    if (isOn)
    {
        if (position < 0)
        {
            ledState1 = ledScale;
        }
        else
        {
            ledState2 = ledScale;
        }
    }

    ledcWrite(ENC_LED_N,  ledState1);
    ledcWrite(ENC_LED_SE, ledState1);
    ledcWrite(ENC_LED_SW, ledState1);

    ledcWrite(ENC_LED_NE, ledState2);
    ledcWrite(ENC_LED_NW, ledState2);
    ledcWrite(ENC_LED_S,  ledState2);

    ledcWrite(ENC_LED_W, 0);
    ledcWrite(ENC_LED_E, 0);
}

void sendToAct()
{
    // Parse settings for transmission ------------------------------------------------------------

    byte outgoingPacket[7], statusByte = 0;
    bool positionNegative = 0;
    int checkWord;

    if (actuator.positionCommand < 0)
    {
        actuator.positionCommand *= -1;
        positionNegative = 1;
    }
    else
        positionNegative = 0;

    statusByte |= actuator.activated;
    statusByte |= actuator.airOut << 1;
    statusByte |= actuator.airIn << 2;
    statusByte |= 0x80; // SYSTEM_TYPE: NimbleStroker

    outgoingPacket[0] = statusByte;
    outgoingPacket[1] = actuator.positionCommand & 0xFF;
    outgoingPacket[2] = actuator.positionCommand >> 8;
    outgoingPacket[2] |= positionNegative << 2;
    outgoingPacket[3] = actuator.forceCommand & 0xFF;
    outgoingPacket[4] = actuator.forceCommand >> 8;

    checkWord = 0;
    for (byte i = 0; i <= 4; i++)
    {
        checkWord += outgoingPacket[i];
    }

    outgoingPacket[5] = checkWord & 0x00FF;
    outgoingPacket[6] = checkWord >> 8;

    for (byte i = 0; i <= 6; i++)
    {
        actSerial.write(outgoingPacket[i]);
    }
}

bool readFromPend()
{
    static byte byteCounter = 0, errorCounter = 0, statusByte = 0;
    static long lastTime = 0, lastPacket = 0;
    static byte incomingPacket[7];
    int checkWord, checkSum;
    bool updated = 0;

    lastPacket = millis() - lastTime;

    if (lastPacket > PACKET_TIMEOUT) // If the last packet was more than the timeout ago, set everything to zero.
    {
        pendant.positionCommand = 0;
        pendant.forceCommand = IDLE_FORCE;
        pendant.present = false;
    }

    while (pendSerial.available()) // Clear pendant incoming serial buffer and fill the incomingPacket array with the first 10 bytes.
    {

        for (byte i = 1; i <= 6; i++) // Shift all bytes in the array to make room for the new one.
            incomingPacket[i - 1] = incomingPacket[i];

        incomingPacket[6] = pendSerial.read(); // put the new byte in the array.

        checkSum = 0; // Reset the checksum before proceeding
        for (byte i = 0; i <= 4; i++)
            checkSum += incomingPacket[i]; // Sum all elements in the array

        checkWord = (incomingPacket[6] << 8) | incomingPacket[5]; // Extract the sent checksum from the array
        if (checkWord == checkSum && checkWord != 0)              // If they match (and aren't zero), update all the variables from the values in the array.
        {
            lastTime = millis(); // Reset the time since the last packet was received.
            statusByte = incomingPacket[0];
            incomingPacket[2] &= 0x07; // Drop the NODE_TYPE designation from this byte.
            incomingPacket[4] &= 0x07; // Drop any random bits from this byte.

            if ((statusByte & 0xE0) == 0x80 && (incomingPacket[4] & 0xF8) == 0) // Verify that the system type and check bits are as expected. If not, don't process the packet.
            {
                pendant.positionCommand = (incomingPacket[2] << 8) | incomingPacket[1];
                if (pendant.positionCommand & 0x0400) // if negative bit is set
                {
                    pendant.positionCommand &= ~(0x0400); // clear negative bit
                    pendant.positionCommand *= -1;        // Set as negative
                }
                pendant.forceCommand = (incomingPacket[4] << 8) | incomingPacket[3];
                pendant.activated = (statusByte & 0x01) ? 1 : 0;
                pendant.airOut = (statusByte & 0x02) ? 1 : 0;
                pendant.airIn = (statusByte & 0x04) ? 1 : 0;
                pendant.present = true;
                updated = 1; // Return 1 since the struct was updated this call.
            }
        }
    }
    return (updated);
}

bool readFromAct()
{
    static byte byteCounter = 0, errorCounter = 0, statusByte = 0;
    static long lastTime = 0, lastPacket = 0;
    static byte incomingPacket[7];
    int checkWord, checkSum;
    bool updated = 0;

    lastPacket = millis() - lastTime;

    if (lastPacket > PACKET_TIMEOUT) // If the last packet was more than the timeout ago, set everything to zero.
        actuator.present = false;

    while (actSerial.available()) // Clear pendant incoming serial buffer and fill the incomingPacket array with the first 10 bytes.
    {
        for (byte i = 1; i <= 6; i++) // Shift all bytes in the array to make room for the new one.
            incomingPacket[i - 1] = incomingPacket[i];

        incomingPacket[6] = actSerial.read(); // put the new byte in the array.

        checkSum = 0; // Reset the checksum before proceeding
        for (byte i = 0; i <= 4; i++)
            checkSum += incomingPacket[i]; // Sum all elements in the array

        checkWord = (incomingPacket[6] << 8) | incomingPacket[5]; // Extract the sent checksum from the array
        if (checkWord == checkSum && checkWord != 0)              // If they match (and aren't zero), update all the variables from the values in the array.
        {
            lastTime = millis(); // Reset the time since the last packet was received.
            statusByte = incomingPacket[0];
            incomingPacket[2] &= 0x07; // Drop the NODE_TYPE designation from this byte.
            incomingPacket[4] &= 0x07; // Drop any random bits from this byte.

            if ((statusByte & 0xE0) == 0x80 && (incomingPacket[4] & 0xF8) == 0) // Verify that the system type and check bits are as expected. If not, don't process the packet.
            {
                actuator.positionFeedback = (incomingPacket[2] << 8) | incomingPacket[1];
                if (actuator.positionFeedback & 0x0400) // if negative bit is set
                {
                    actuator.positionFeedback &= ~(0x0400); // clear negative bit
                    actuator.positionFeedback *= -1;        // Set as negative
                }
                actuator.forceFeedback = (incomingPacket[4] << 8) | incomingPacket[3];
                if (actuator.forceFeedback & 0x0400) // if negative bit is set
                {
                    actuator.forceFeedback &= ~(0x0400); // clear negative bit
                    actuator.forceFeedback *= -1;        // Set as negative
                }
                actuator.activated = (statusByte & 0x01) ? 1 : 0;
                actuator.sensorFault = (statusByte & 0x02) ? 1 : 0;
                actuator.tempLimiting = (statusByte & 0x04) ? 1 : 0;
                actuator.present = true;
                updated = 1; // Return 1 since the struct was updated this call.
            }
        }
    }
    return (updated);
}
